src/wasm/mod.rs

```diff
@@ -0,0 +1,233 @@
+            _ => {
+                return Err(wasmi::Error::Instantiation(
+                    "required memory export".to_owned(),
+                ))
+            }
+        };
+
+        Ok(Host {
+            module: instance.clone(),
+            memory,
+        })
+    }
+}
+
+macro_rules! host_funcs {
+    ($(
+        fn $module:literal :: $name:literal ($($p:ident : $t:ident),*) $( -> $rt:ident)?
+            as $variant:ident impl $method:path;
+    )*) => {
+        #[repr(usize)]
+        #[derive(Copy, Clone, Debug, Eq, PartialEq)]
+        enum HostFunc {
+            $($variant),*
+        }
+
+        impl HostFunc {
+            fn resolve_func(
+                module_name: &str,
+                field_name: &str,
+                _signature: &wasmi::Signature,
+            ) -> Result<HostFunc, wasmi::Error> {
+                match (module_name, field_name) {
+                    $(($module, $name) => Ok(HostFunc::$variant),)*
+                    _ => Err(wasmi::Error::Instantiation("unresolved func import".to_owned()))
+                }
+            }
+
+            fn signature(self) -> wasmi::Signature {
+                match self {
+                    $(
+                        HostFunc::$variant => wasmi::Signature::new(
+                            &[$(<$t as WasmPrimitive>::TYPE),*][..],
+                            option_helper!(Some $(<$rt as WasmPrimitive>::TYPE)?),
+                        )
+                    ),*
+                }
+            }
+
+            fn func_ref(self) -> wasmi::FuncRef {
+                wasmi::FuncInstance::alloc_host(self.signature(), self as usize)
+            }
+
+            fn module_name(self) -> &'static str {
+                match self {
+                    $(HostFunc::$variant => $module),*
+                }
+            }
+
+            fn field_name(self) -> &'static str {
+                match self {
+                    $(HostFunc::$variant => $name),*
+                }
+            }
+        }
+
+        impl TryFrom<usize> for HostFunc {
+            type Error = wasmi::Trap;
+            fn try_from(x: usize) -> Result<Self, Self::Error> {
+                $(
+                    if x == (HostFunc::$variant as usize) {
+                        return Ok(HostFunc::$variant);
+                    }
+                )*
+                Err(wasmi::TrapKind::UnexpectedSignature.into())
+            }
+        }
+
+        impl fmt::Display for HostFunc {
+            fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+                write!(f, "{}::{}", self.module_name(), self.field_name())
+            }
+        }
+
+        impl wasmi::Externals for Host {
+            fn invoke_index(
+                &mut self,
+                index: usize,
+                args: wasmi::RuntimeArgs,
+            ) -> Result<Option<wasmi::RuntimeValue>, wasmi::Trap> {
+                let span = tracing::trace_span!("invoke_index", index, ?args);
```

Comment:
TIOLI: seems like we could just slap a `#[tracing::instrument]` on the generated function?

Reply:
This is the method which I was poking you on Discord about yesterday which breaks when annotated with `#[tracing::instrument]` due to rust messing up spans.

Reply:
ah gotcha. it would be nice to have a tracing issue for that (even though it's a compiler error), just so we can reference the related issue against the compiler?